### PR TITLE
WIP: Use Abort marker trait to audit OOM panics

### LINKED ISSUE #15: Fix `unchecked_unwraped` in `Box`

### PR DESCRIPTION
**Pull request by Ericson2314**
This avoids the non-untuitive behavior of the `try_*` methods from
`Error = !`, while still giving us the ability to audit panicks.
Perfect!

Credit to @TimDiekmann for the idea.

`AbortAlloc` is put back as an private implementation detail to make the
unchecked unwrapps in `boxed` safe. A simpler solution would be just get
in the habit of returning the layout and allocation error, as in the
general case the caller has now idea which allocation deep in the callee
failed. Not returning the layout only makes sense for the allocator
trait itself, when the caller must know the layout.

WIP because there are some doc test compile failures I don't have time
to sort out at this moment.

closes #15

### CODE REVIEW

File: src/vec.rs [hunk 1]
```diff
@@ -844,13 +844,9 @@ impl<T, A: DeallocRef> Vec<T, A> {
...
     where
-        A: ReallocRef,
+        A: ReallocRef + Abort,
     {
-        match self.try_into_boxed_slice() {
-            Ok(vec) => vec,
-            Err(CollectionAllocErr::CapacityOverflow) => capacity_overflow(),
-            Err(CollectionAllocErr::AllocError { layout, .. }) => handle_alloc_error(layout.into()),
-        }
+        handle_collection_error_audited(self.try_into_boxed_slice())
```
Comment (TimDiekmann) [inline src/vec.rs:849]: This does not panic, but aborts on `CapacityOverflow` now.

Comment (Ericson2314) [reply src/vec.rs:849]: Really? Scrolling down in my phone (so easy to miss things, take with grain of salt) it looks like I just moved the panic into this `handle_*` function. Did I not?

Comment (TimDiekmann) [reply src/vec.rs:849]: `handle_alloc_error` does not `panic!` but aborts the program

Comment (TimDiekmann) [reply src/vec.rs:849]: Actually it calls a handler, which can me modified, so a capacity overflow has to be handled differently anyways.

File: src/lib.rs [hunk 2]
```diff
@@ -67,7 +67,8 @@
     specialization,
     trusted_len,
     unsized_locals,
-    fn_traits
+    fn_traits,
+    exhaustive_patterns
```
Comment (TimDiekmann) [inline src/lib.rs:71]: As far as I can see, there is no rustc version, which supports clippy and rustfmt with the never_type stabilized. We might want to add `never_type` as feature until this is resolved.
```suggestion
    exhaustive_patterns,
    never_type
```

File: src/alloc/abort.rs [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    ReallocRef,
+    System,
+};
+use core::ptr::NonNull;
+
+/// Indicates that the allocator is allowed to be used where we panic on OOM.
+pub trait Abort: AllocRef {}
+
+impl Abort for Global {}
+impl Abort for System {}
```
Comment (TimDiekmann) [inline src/alloc/abort.rs:21]: `System` needs std support:

```suggestion
#[cfg(feature = "std")]
impl Abort for std::alloc::System {}
```

File: src/alloc/abort.rs [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+
+use crate::alloc::{
+    AllocRef,
+    BuildAllocRef,
+    DeallocRef,
+    Global,
+    NonZeroLayout,
+    NonZeroUsize,
+    ReallocRef,
+    System,
```
Comment (TimDiekmann) [inline src/alloc/abort.rs:13]:
```suggestion
```

File: src/alloc/abort.rs [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+};
+use core::ptr::NonNull;
+
+/// Indicates that the allocator is allowed to be used where we panic on OOM.
+pub trait Abort: AllocRef {}
```
Comment (TimDiekmann) [inline src/alloc/abort.rs:18]: Is the bound really needed or just a hint to the user, that `Abort` without `AllocRef` does not make much sense? If the latter, I'd remove the bound and require both traits on the methods to be explicit. It also requires one click less for unexperienced user in the documentation.

File: src/alloc/mod.rs [hunk 6]
```diff
@@ -9,10 +10,12 @@ use core::{
     num::NonZeroUsize,
     ptr::{self, NonNull},
 };
-pub use liballoc::alloc::{handle_alloc_error, Layout};
+use liballoc::alloc::Layout;
```
Comment (TimDiekmann) [inline src/alloc/mod.rs:13]: Let's reexport everything, which has already been stabilized.
```suggestion
pub use liballoc::alloc::Layout;
```

**Review by TimDiekmann — CHANGES_REQUESTED** [review]

File: src/alloc/mod.rs (hunk 6 above)
Comment (Ericson2314) [reply src/alloc/mod.rs:13]: Oh, sorry! Meant to just not `pub use` `handle_alloc_error`.

**Comment by Ericson2314** [comment]
All of these suggestions look good to me. Will try to get them later today.

**Comment by TimDiekmann** [comment]
Were you able to make any progression on this so far? If not, I'll merge this later and fix the tests :slightly_smiling_face:

File: src/lib.rs (hunk 2 above)
Comment (TimDiekmann) [reply src/lib.rs:71]: Clippy works with todays nightly

**Comment by Ericson2314** [comment]
Thanks for picking this up!

Yeah whenever I make a time commitment like that on github, I invariably mess it up. And, lo and behold, I had more things to do at work, and little energy to program at home this week.

**Comment by TimDiekmann** [comment]
No problem :)

I removed `AbortAlloc` as it only was used in two places in `Box`. I'm still pretty happy about this approach!

**Pull request merged**